test/Effect.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE DeriveFunctor, ExistentialQuantification, FlexibleInstances, GeneralizedNewtypeDeriving, MultiParamTypeClasses, StandaloneDeriving, TypeFamilies, TypeOperators, UndecidableInstances #-}
+module Effect
+( tests
+) where
+
+import Control.Algebra
+import Control.Carrier.NonDet.Church
+import Control.Carrier.State.Strict
+import Control.Monad.Trans.Class
+import Data.Functor.Identity
+import Data.Monoid (Sum(..))
+import Prelude hiding (all)
+import Test.Tasty
+import Test.Tasty.HUnit
+
+tests :: TestTree
+tests = testGroup "Effect"
+  [ testCase "allows effects to constrain the state functor" $ do
+    let bs = run . runAll . runState (sum 0) $ do
+          s <- pure "hello" <|> pure "world"
+          bs <- all (foldMapA (\ (c, i) -> c <$ put (sum i)) (zip s [1..]))
+          bs <$ put (sum 1) <|> pure bs
+    bs @?= [(Sum 1, "hello"), (Sum 15, "hello"), (Sum 1, "world"), (Sum 15, "world")]
+  ] where
+  sum i = Sum (i :: Int)
+
+
+-- | An effect to return the results of all of the nondeterministic branches computed within its scope.
+data All m k = forall a . All (m a) ([a] -> m k)
+
+deriving instance Functor m => Functor (All m)
+
+-- | Note that since the continuation takes @[a]@, our 'Effect' instance will have to produce a continuation taking @[f a]@ given one taking @[a]@, for which purpose we use 'sequenceA'. That necessitates an 'Applicative' instance for @f@, so we specialize 'CanHandle' for this effect to strengthen the constraint.
+--
+-- Intuitively, we can think of this as saying that if there are 'Control.Effect.State.State' effects present in @m@, collecting all of their results—joining the branches—requires us to merge the states monoidally; if there are 'Control.Effect.Error.Error' effects, it requires that they’ve all succeeded.
+instance Effect All where
+  type CanHandle All f = Applicative f
+  handle ctx hdl (All m k) = All (hdl (m <$ ctx)) (hdl . fmap k . sequenceA)
+
+all :: Has All sig m => m a -> m [a]
+all m = send (All m pure)
```

Comment:
I am not proposing to include this in the library until such time as we have built up confidence in it in e.g. `semantic`.

That said, this is _precisely_ the operation I talked with @zenzike about at ICFP 2018, when I had finally built up an intuition for _why_ it required `Applicative` contexts, and so I am extremely excited that it’s now possible to express in `fused-effects`!